library/ssl_cache.c

```diff
@@ -137,158 +124,191 @@ int mbedtls_ssl_cache_get( void *data, mbedtls_ssl_session *session )
     while( cur != NULL )
     {
         count++;
 
+        if( session_id_len == cur->session_id_len &&
+            memcmp( session_id, cur->session_id, cur->session_id_len ) == 0 )
+        {
+            goto found;
+        }
+
+        prv = cur;
+        cur = cur->next;
+    }
+
+    /* Check 2: Is there an outdated entry in the cache?
+     *
+     * If so, overwrite it.
+     *
+     * If not, remember the oldest entry in `old` for later.
+     */
+
 #if defined(MBEDTLS_HAVE_TIME)
+    while( cur != NULL )
+    {
         if( cache->timeout != 0 &&
             (int) ( t - cur->timestamp ) > cache->timeout )
         {
-            cur->timestamp = t;
-            break; /* expired, reuse this slot, update timestamp */
+            goto found;
         }
-#endif
 
-        if( memcmp( session->id, cur->session.id, cur->session.id_len ) == 0 )
-            break; /* client reconnected, keep timestamp for session id */
-
-#if defined(MBEDTLS_HAVE_TIME)
         if( oldest == 0 || cur->timestamp < oldest )
         {
             oldest = cur->timestamp;
             old = cur;
         }
-#endif
 
         prv = cur;
         cur = cur->next;
     }
+#endif /* MBEDTLS_HAVE_TIME */
 
-    if( cur == NULL )
+    /* Check 3: Is there free space in the cache? */
+
+    if( count < cache->max_entries )
     {
-#if defined(MBEDTLS_HAVE_TIME)
-        /*
-         * Reuse oldest entry if max_entries reached
-         */
-        if( count >= cache->max_entries )
-        {
-            if( old == NULL )
-            {
-                ret = 1;
-                goto exit;
-            }
+        /* Create new entry */
+        cur = mbedtls_calloc( 1, sizeof(mbedtls_ssl_cache_entry) );
+        if( cur == NULL )
+            return( 1 );
+
+        /* Append to the end of the linked list. */
+        if( prv == NULL )
+            cache->chain = cur;
+        else
+            prv->next = cur;
 
-            cur = old;
-        }
+        goto found;
+    }
+
+    /* Last resort: The cache is full and doesn't contain any outdated
+     * elements. In this case, we evict the oldest one, judged by timestamp
+     * (if present) or cache-order. */
+
+#if defined(MBEDTLS_HAVE_TIME)
+    if( old == NULL )
+    {
+        /* This should only happen on an ill-configured cache
+         * with max_entries == 0. */
+        return( 1 );
+    }
 #else /* MBEDTLS_HAVE_TIME */
-        /*
-         * Reuse first entry in chain if max_entries reached,
-         * but move to last place
-         */
-        if( count >= cache->max_entries )
-        {
-            if( cache->chain == NULL )
-            {
-                ret = 1;
-                goto exit;
-            }
-
-            cur = cache->chain;
-            cache->chain = cur->next;
-            cur->next = NULL;
-            prv->next = cur;
-        }
+    /* Reuse first entry in chain, but move to last place. */
+    if( cache->chain == NULL )
+        return( 1 );
+
+    old = cache->chain;
```

Comment:
Travis is complaining that old is undeclared here.

Reply:
Rightfully so. I've changed the code to use always use `old` as the oldest entry in the cache, where 'oldest' is judged by timestamp if `MBEDTLS_HAVE_TIME`, _or_ by cache-order. However, I had forgotten to remove the `MBEDTLS_HAVE_TIME`-guard around the declaration. Thank you!

Reply:
https://github.com/ARMmbed/mbedtls/pull/4347/commits/006f2cce2eb1b73fa966fe52e75089dd5ec34912